codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttEncoder.java

```diff
@@ -284,35 +340,72 @@ private static ByteBuf encodeUnsubscribeMessage(
             MqttSubAckMessage message) {
-        int variableHeaderBufferSize = 2;
+        MqttVersion mqttVersion = MqttCodecUtil.getMqttVersion(ctx);
+        ByteBuf propertiesBuf = encodePropertiesIfNeeded(mqttVersion,
+                byteBufAllocator,
+                message.idAndPropertiesVariableHeader().properties());
+        int variableHeaderBufferSize = 2 + propertiesBuf.readableBytes();
         int payloadBufferSize = message.payload().grantedQoSLevels().size();
         int variablePartSize = variableHeaderBufferSize + payloadBufferSize;
         int fixedHeaderBufferSize = 1 + getVariableLengthInt(variablePartSize);
         ByteBuf buf = byteBufAllocator.buffer(fixedHeaderBufferSize + variablePartSize);
         buf.writeByte(getFixedHeaderByte1(message.fixedHeader()));
         writeVariableLengthInt(buf, variablePartSize);
         buf.writeShort(message.variableHeader().messageId());
+        buf.writeBytes(propertiesBuf);
         for (int qos : message.payload().grantedQoSLevels()) {
             buf.writeByte(qos);
         }
 
         return buf;
     }
 
+    private static ByteBuf encodeUnsubAckMessage(
+            ChannelHandlerContext ctx,
+            ByteBufAllocator byteBufAllocator,
+            MqttUnsubAckMessage message) {
+        MqttVersion mqttVersion = MqttCodecUtil.getMqttVersion(ctx);
+        ByteBuf propertiesBuf = encodePropertiesIfNeeded(mqttVersion,
+                byteBufAllocator,
+                message.idAndPropertiesVariableHeader().properties());
+        int variableHeaderBufferSize = 2 + propertiesBuf.readableBytes();
+        int payloadBufferSize = message.payload().unsubscribeReasonCodes().size();
+        int variablePartSize = variableHeaderBufferSize + payloadBufferSize;
+        int fixedHeaderBufferSize = 1 + getVariableLengthInt(variablePartSize);
+        ByteBuf buf = byteBufAllocator.buffer(fixedHeaderBufferSize + variablePartSize);
+        buf.writeByte(getFixedHeaderByte1(message.fixedHeader()));
+        writeVariableLengthInt(buf, variablePartSize);
+        buf.writeShort(message.variableHeader().messageId());
+        buf.writeBytes(propertiesBuf);
+
+        for (Short reasonCode : message.payload().unsubscribeReasonCodes()) {
+            buf.writeByte(reasonCode);
+        }
+
+        return buf;
+    }
+
     private static ByteBuf encodePublishMessage(
+            ChannelHandlerContext ctx,
             ByteBufAllocator byteBufAllocator,
             MqttPublishMessage message) {
+        MqttVersion mqttVersion = MqttCodecUtil.getMqttVersion(ctx);
         MqttFixedHeader mqttFixedHeader = message.fixedHeader();
         MqttPublishVariableHeader variableHeader = message.variableHeader();
         ByteBuf payload = message.payload().duplicate();
 
         String topicName = variableHeader.topicName();
         byte[] topicNameBytes = encodeStringUtf8(topicName);
 
+        ByteBuf propertiesBuf = encodePropertiesIfNeeded(mqttVersion,
```

Comment:
you will need to call `propertiesBuf.release()` later as you not transfer ownership